pennylane_qiskit/qiskit_device.py

```diff
@@ -12,124 +12,204 @@
+    """
+    # Code to acquire session:
+    existing_session = device._session
+    session = Session(backend=device.backend)
+    device._session = session
     try:
-        return backend.name()  # BackendV1
-    except TypeError:  # pragma: no cover
-        return backend.name  # BackendV2
+        yield session
+    finally:
+        # Code to release session:
+        session.close()
+        device._session = existing_session
+
+
+def accepted_sample_measurement(m: qml.measurements.MeasurementProcess) -> bool:
+    """Specifies whether or not a measurement is accepted when sampling."""
+
+    return isinstance(
+        m,
+        (
+            qml.measurements.SampleMeasurement,
+            qml.measurements.ClassicalShadowMP,
+            qml.measurements.ShadowExpvalMP,
+        ),
+    )
 
 
-class QiskitDevice(QubitDevice, abc.ABC):
-    r"""Abstract Qiskit device for PennyLane.
+@transform
+def split_execution_types(
+    tape: qml.tape.QuantumTape,
+) -> (Sequence[qml.tape.QuantumTape], Callable):
+    """Split into separate tapes based on measurement type. Counts and sample-based measurements
+    will use the Qiskit Sampler. ExpectationValue and Variance will use the Estimator, except
+    when the measured observable does not have a `pauli_rep`. In that case, the Sampler will be
+    used, and the raw samples will be processed to give an expectation value."""
+    estimator = []
+    sampler = []
+
+    for i, mp in enumerate(tape.measurements):
+        if isinstance(mp, (ExpectationMP, VarianceMP)):
+            if mp.obs.pauli_rep:
+                estimator.append((mp, i))
+            else:
+                warnings.warn(
+                    f"The observable measured {mp.obs} does not have a `pauli_rep` "
+                    "and will be run without using the Estimator primitive. Instead, "
+                    "raw samples from the Sampler will be used."
+                )
+                sampler.append((mp, i))
+        else:
+            sampler.append((mp, i))
+
+    order_indices = [[i for mp, i in group] for group in [estimator, sampler]]
+
+    tapes = []
+    if estimator:
+        tapes.extend(
+            [
+                qml.tape.QuantumScript(
+                    tape.operations,
+                    measurements=[mp for mp, i in estimator],
+                    shots=tape.shots,
+                )
+            ]
+        )
+    if sampler:
+        tapes.extend(
+            [
+                qml.tape.QuantumScript(
+                    tape.operations,
+                    measurements=[mp for mp, i in sampler],
+                    shots=tape.shots,
+                )
+            ]
+        )
+
+    def reorder_fn(res):
+        """re-order the output to the original shape and order"""
+
+        flattened_indices = [i for group in order_indices for i in group]
+        flattened_results = [r for group in res for r in group]
+
+        result = dict(zip(flattened_indices, flattened_results))
```

Comment:
We should `assert` the length before zipping so we can ensure we don't miss anything. You can skip codecov check for the condition in case you don't have a unit test for it.